clang/lib/DPCT/APINamesCooperativeGroups.inc

```diff
@@ -541,6 +541,95 @@ CASE_FACTORY_ENTRY(
                                       Diagnostics::API_NOT_MIGRATED,
                                       ARG("shfl_up"))))
 
+/*
+  --- X.shfl() ---
+  cg::thread_block_tile<32>::shfl(v, src_rank)
+  => sycl::select_from_group(item_ct1.get_sub_group(), v, src_rank)
+
+  cg::thread_block_tile<N>::shfl(v, src_rank) if N in {1,2,4,8,16}
```

Comment:
This PR in general looks troublesome to me for a few reasons. This code is one example.

It seems to be mapping to a masked subgroup shuffle to a non-masked variant of `dpct::select_from_sub_group`. I don't see how this can do anything other than break code.

In general I don't know why you don't use the non-uniform groups that is already implemented in dpc++ and will be moved to experimental here: https://github.com/intel/llvm/pull/11535

If you want to introduced syclomatic mappings to `cg::thread_block_tile<N>::shfl` then we can add `fixed_size_group::shfl` to this extension.
I understand that `cg::thread_block_tile<N>::shfl` is not yet part of the non-uniform groups extension, and that sometimes you want a temporary solution. However if that temporary solution is not a valid mapping then perhaps you can wait until a valid mapping exists?

Now a better mapping would be to the masked shuffle functions that I implement here: https://github.com/oneapi-src/SYCLomatic/pull/1342
However even this I think would not generally result in legal code, since it removes the concept of the thread_block_tile, so the mapping would only be legal if the rest of the program is still valid after throwing away the thread_block_tile/ concept. Achieving this without a mapping to the corresponding group class in dpc++ (fixed_size_group) looks troublesome.

cc @Pennycook

Reply:
Hi jackAKirk, 

After the https://github.com/intel/llvm/pull/11535 was merged into the repository, the fixed_size_group in the extension is not supported in the sycl::select_from_group. We have to wait the algorithm implementation. Therefore, we prefer to continue using the helper function in the code, as the helper is a general-purpose code that can be used on any SYCL-supported device. The experimental implementation is specific to Intel, and we can transition to the extension when it is incorporated into the SYCL specification.


SYCLomatic will adopt the experimental feature if its implementation is mature or has a similar behavior as the migration API. However, the experimental feature may be deprecated at some point in the future, so we prefer implement it in the helper funciton.

Reply:
After intel/llvm#11535 is mature enough,  we can further update the migration logic. Prefer to merge the this PR first.

Reply:
I think that you've missed the point I was making. The translation takes code that does one thing and turns it into code that does a different thing. I don't understand the value in such a miss-"translation". It would be better to not (miss)"translate" it at all surely.